File: LC576.py
import pprint


def findPaths(m, n, maxMove, startRow, startColumn):
    arr = [[[-1 for i in range(m)] for j in range(n)] for k in range(maxMove + 1)]
    pprint(arr)
    return recurse(arr, m, n, startRow, startColumn, maxMove)


def recurse(arr, m, n, startRow, startColumn, moves):
    # we got outside from the previous move
    if startRow < 0 or startColumn < 0 or startRow >= m or startColumn >= n:
        return 1
    # we ran out of moves
    if moves == 0:
        return 0
    # optimization to not repeat work
    if arr[moves][startColumn][startRow] >= 0:
        return arr[moves][startColumn][startRow]
    # recursive step where we step in every direction
    arr[moves][startColumn][startRow] = (
        recurse(arr, m, n, startRow + 1, startColumn, moves - 1)
        + recurse(
            arr,
            m,
            n,
            startRow - 1,
            startColumn,
            moves - 1,
        )
        + recurse(arr, m, n, startRow, startColumn + 1, moves - 1)
        + recurse(arr, m, n, startRow, startColumn - 1, moves - 1)
    )
    return arr[moves][startColumn][startRow]

File: JavaSolutions/LC576.java
int[][][] numPaths;

    public int findPaths(int m, int n, int maxMove, int startRow, int startColumn) {
        numPaths= new int[m][n][maxMove+1];

        for(int[][] arr2d:numPaths){
            for(int[] arr:arr2d){
                Arrays.fill(arr,-1);
            }
        }
        
        return recurse(m, n, startRow, startColumn, maxMove);

    }

    public int recurse(int m, int n, int startRow, int startColumn, int numMoves) {
        if (startRow < 0||startColumn == n || startColumn < 0||startRow == m ) return 1;
        if(numMoves==0) return 0;
        if(numPaths[startRow][startColumn][numMoves]>=0) return numPaths[startRow][startColumn][numMoves];

        numPaths[startRow][startColumn][numMoves] = ((recurse(m, n, startRow - 1, startColumn, numMoves - 1)+recurse(m, n, startRow + 1, startColumn, numMoves - 1))% 1000000007+(recurse(m, n, startRow, startColumn + 1, numMoves - 1)+recurse(m, n, startRow, startColumn - 1, numMoves - 1))% 1000000007)% 1000000007;
        return numPaths[startRow][startColumn][numMoves];
    }
